### ISSUE TITLE
Error when entering Visual Block Mode if using the whole lua code inside init.vim

### ISSUE THREAD

**Issue by xfzv** [Avimitin/nerd-galaxyline#4]
Hi,

First, thank you for sharing this. My lua skills are very limited and it was easier for to me use your bar as a base and then customize to my liking. When using the plugin with vim-plug, no issue.

But I prefer to use the whole [`nerd-galaxyline.lua` content](https://github.com/Avimitin/nerd-galaxyline/blob/master/lua/nerd-galaxyline.lua) inside my `init.vim` file (since I customized it) with

```vim
lua <<EOF
[...]
EOF
```

Everything works well, except that I'm getting an error when entering visual block mode with Ctrl+Alt+v:

```vim
[string ":lua"]:206: attempt to concatenate a nil value funct...tin#18 [string ":lua"]:206: attempt to concatenate a nil value
E15: Invalid expression: luaeval('require("galaxyline").component_decorator')("ViMode")
Press ENTER or type command to continue
```

The error also occurs with the default [`nerd-galaxyline.lua` content](https://github.com/Avimitin/nerd-galaxyline/blob/master/lua/nerd-galaxyline.lua) (without any modification).

Commenting out this whole block solves the issue:

https://github.com/Avimitin/nerd-galaxyline/blob/a0949262a3ea9421a7cc43068a2d7c3dfdd3fcba/lua/nerd-galaxyline.lua#L158-L211

`% nvim --version`:

```log
NVIM v0.5.0
Build type: Gentoo
LuaJIT 2.0.5
Compilation: /usr/bin/x86_64-pc-linux-gnu-gcc -march=native -O2 -pipe -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/var/tmp/portage/app-editors/neovim-0.5.0/work/neovim-0.5.0_build/config -I/var/tmp/portage/app-editors/neovim-0.5.0/work/neovim-0.5.0/src -I/usr/include -I/var/tmp/portage/app-editors/neovim-0.5.0/work/neovim-0.5.0_build/src/nvim/auto -I/var/tmp/portage/app-editors/neovim-0.5.0/work/neovim-0.5.0_build/include
Compiled by portage@localhost

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "/etc/vim/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

Any idea?

Thanks.

**Comment by Avimitin** [comment]
This might be some escape character error. I will try to figure it out.

**Comment by Avimitin** [comment]
Can you send your init.vim file?

**Comment by Avimitin** [comment]
Also, try to comment out the 206 lines, which tries to highlight the status bar. See if the status bar crash again.

**Comment by xfzv** [comment]
Indeed, commenting out only

https://github.com/Avimitin/nerd-galaxyline/blob/a0949262a3ea9421a7cc43068a2d7c3dfdd3fcba/lua/nerd-galaxyline.lua#L206

works, but it also disables the whole Vim mode section:
![image](https://user-images.githubusercontent.com/78810647/130637917-d8fb033b-94f1-4501-94d1-13fceaaf2758.png)

<details>
  <summary>Here's my stripped down init.vim so that you can reproduce the issue:</summary>

```vim
call plug#begin("~/.config/nvim/plugged")
Plug 'glepnir/galaxyline.nvim' , {'branch': 'main'}
Plug 'kyazdani42/nvim-web-devicons'
call plug#end()

lua <<EOF
local present1, gl = pcall(require, 'galaxyline')
if not present1 then
    print("galxyline not found")
    return
end

local gls = gl.section

gl.short_line_list = {
    'LuaTree',
    'vista',
    'dbui',
    'startify',
    'term',
    'nerdtree',
    'fugitive',
    'fugitiveblame',
    'plug'
}

-- VistaPlugin = extension.vista_nearest

local colors = {
    bg       = '#C8CCD4',
    line_bg  = '#1E1E20',
    fg       = '#8FBCBB',
    fg_green = '#98C379',

    yellow   = '#E5C07B',
    cyan     = '#56B6C2',
    darkblue = '#83A598',
    green    = '#98C379',
    orange   = '#D19A66',
    purple   = '#C678DD',
    magenta  = '#C858E9',
    blue     = '#61AFEF',
    red      = '#BE5046',
}

local use_coc = false
if vim.g.nerd_galaxyline_lsp == 'coc' then
    use_coc = true
end

local function lsp_status(status)
    local shorter_stat = ''
    for match in string.gmatch(status, "[^%s]+")  do
        local err_warn = string.find(match, "^[WE]%d+", 0)
        if not err_warn then
            shorter_stat = shorter_stat .. ' ' .. match
        end
    end
    return shorter_stat
end

local function get_coc_lsp()
  local status = vim.fn['coc#status']()
  if not status or status == '' then
      return ''
  end
  return lsp_status(status)
end

local function get_diagnostic_info()
  if vim.fn.exists('*coc#rpc#start_server') == 1 then
    return get_coc_lsp()
    end
  return ''
end

local function get_current_func()
  local has_func, func_name = pcall(vim.fn.nvim_buf_get_var,0,'coc_current_function')
  if not has_func then return end
      return func_name
  end

local function get_function_info()
  if vim.fn.exists('*coc#rpc#start_server') == 1 then
    return get_current_func()
    end
  return ''
end

local function trailing_whitespace()
    local trail = vim.fn.search("\\s$", "nw")
    if trail ~= 0 then
        return ' '
    else
        return nil
    end
end

CocStatus = get_diagnostic_info
CocFunc = get_function_info
TrailingWhiteSpace = trailing_whitespace

local function has_file_type()
    local f_type = vim.bo.filetype
    if not f_type or f_type == '' then
        return false
    end
    return true
end

local buffer_not_empty = function()
  if vim.fn.empty(vim.fn.expand('%:t')) ~= 1 then
    return true
  end
  return false
end

-- insert_left insert item at the left panel
local function insert_left(element)
  table.insert(gls.left, element)
end

-- insert_blank_line_at_left insert blank line with
-- line_bg color.
local function insert_blank_line_at_left()
insert_left {
  Space = {
    provider = function () return ' ' end,
    highlight = {colors.line_bg,colors.line_bg}
  }
}
end

-- insert_right insert given item into galaxyline.right
local function insert_right(element)
  table.insert(gls.right, element)
end

-- insert_blank_line_at_left insert blank line with
-- line_bg color.
local function insert_blank_line_at_right()
insert_right {
  Space = {
    provider = function () return ' ' end,
    highlight = {colors.line_bg,colors.line_bg}
  }
}
end

-----------------------------------------------------
----------------- start insert ----------------------
-----------------------------------------------------

--{ mode panel start
insert_left{
  Start = {
    provider = function() return ' ' end,
    highlight = {colors.line_bg,}
  }
}

insert_blank_line_at_left()

insert_left{
  ViMode = {
    icon = function()
        local icons = {
            n      = ' ',
            i      = ' ',
            c      = 'ﲵ ',
            V      = ' ',
            [''] = ' ',
            v      = ' ',
            C      = 'ﲵ ',
            R      = '﯒ ',
            t      = ' ',
        }
        return icons[vim.fn.mode()]
      end,
    provider = function()
      -- auto change color according the vim mode
      local alias = {
          n      = 'N',
          i      = 'I',
          c      = 'C',
          V      = 'VL',
          [''] = 'V',
          v      = 'V',
          C      = 'C',
          ['r?'] = ':CONFIRM',
          rm     = '--MORE',
          R      = 'R',
          Rv     = 'R&V',
          s      = 'S',
          S      = 'S',
          ['r']  = 'HIT-ENTER',
          [''] = 'SELECT',
          t      = 'T',
          ['!']  = 'SH',
      }
      local mode_color = {
        n = colors.yellow,      i = colors.green,   v=colors.blue,
        [''] = colors.blue,   V=colors.blue,      c = colors.magenta,
        no = colors.red,        s = colors.orange,  S=colors.orange,
        [''] = colors.orange, ic = colors.yellow, R = colors.purple,
        Rv = colors.purple,     cv = colors.red,    ce=colors.red,
        r = colors.cyan,        rm = colors.cyan,   ['r?'] = colors.cyan,
        ['!'] = colors.red,     t = colors.red
      }

      local vim_mode = vim.fn.mode()
      vim.api.nvim_command('hi GalaxyViMode guifg='..mode_color[vim_mode])
      return alias[vim_mode]
    end,
    highlight = {colors.line_bg, colors.line_bg},
  },
}

insert_blank_line_at_left()

insert_left{
  Separa = {
    provider = function() return ' ' end,
    highlight = {colors.line_bg, },
  }
}

--mode panel end}

-- {information panel start
insert_left{
  Start = {
    provider = function() return ' ' end,
    highlight = {colors.line_bg,}
  }
}

insert_left{
  FileIcon = {
    provider = 'FileIcon',
    condition = buffer_not_empty,
    highlight = {require('galaxyline.provider_fileinfo').get_file_icon_color,colors.line_bg},
  },
}

insert_left{
  FileName = {
    provider = function()
      return vim.fn.expand("%:F")
    end,
    condition = function() return buffer_not_empty and has_file_type() end,
    highlight = {colors.bg,colors.line_bg}
  }
}

insert_blank_line_at_left()

insert_left {
  GitIcon = {
    provider = function() return '  ' end,
    condition = require('galaxyline.provider_vcs').check_git_workspace,
    highlight = {colors.orange,colors.line_bg},
  }
}

insert_left {
  GitBranch = {
    provider = 'GitBranch',
    condition = require('galaxyline.provider_vcs').check_git_workspace,
    highlight = {'#8FBCBB',colors.line_bg,'bold'},
  }
}

insert_blank_line_at_left()

local checkwidth = function()
  local squeeze_width  = vim.fn.winwidth(0) / 2
  if squeeze_width > 40 then
    return true
  end
  return false
end

insert_left {
  DiffAdd = {
    provider = 'DiffAdd',
    condition = checkwidth,
    icon = '  ',
    highlight = {colors.green,colors.line_bg},
  }
}

insert_left {
  DiffModified = {
    provider = 'DiffModified',
    condition = checkwidth,
    icon = '  ',
    highlight = {colors.orange,colors.line_bg},
  }
}

insert_left {
  DiffRemove = {
    provider = 'DiffRemove',
    condition = checkwidth,
    icon = '  ',
    highlight = {colors.red,colors.line_bg},
  }
}

insert_left {
    TrailingWhiteSpace = {
     provider = TrailingWhiteSpace,
     icon = '  ',
     highlight = {colors.yellow,colors.line_bg},
    }
}

insert_left {
  DiagnosticError = {
    provider = 'DiagnosticError',
    icon = '  ',
    highlight = {colors.red,colors.line_bg}
  }
}

insert_left {
  DiagnosticWarn = {
    provider = 'DiagnosticWarn',
    icon = '  ',
    highlight = {colors.yellow,colors.line_bg},
  }
}

insert_left {
    CocStatus = {
     provider = CocStatus,
     highlight = {colors.green,colors.line_bg},
     icon = '  ',
         condition = use_coc,
    }
}

insert_left {
    CocStatus = {
     provider = 'DiagnosticInfo',
     highlight = {colors.green,colors.line_bg},
     icon = '  ',
         condition = function() return checkwidth() and not use_coc end,
    }
}

insert_left {
  CocFunc = {
    provider = CocFunc,
    icon = ' λ ',
    highlight = {colors.yellow,colors.line_bg},
        condition = use_coc,
  }
}

insert_left {
    DiagnosticHint = {
     provider = 'DiagnosticHint',
     condition = function() return checkwidth() and not use_coc end,
     highlight = {colors.white,colors.line_bg},
     icon = '  ',
    }
}

insert_left{
  Separa = {
    provider = function() return ' ' end,
    highlight = {colors.line_bg, },
  }
}
-- left information panel end}

insert_right{
  Start = {
    provider = function() return ' ' end,
    highlight = {colors.line_bg,}
  }
}

insert_blank_line_at_right()

-- insert_right{
  -- FileFormat = {
    -- provider = 'FileFormat',
        -- condition = checkwidth,
    -- highlight = {colors.fg,colors.line_bg,'bold'},
  -- }
-- }

insert_blank_line_at_right()

insert_right{
  FileType = {
    provider = function() return vim.bo.filetype end,
    separator = ' ',
    separator_highlight = {colors.red,colors.line_bg},
    highlight = {colors.red,colors.line_bg,'bold'},
        condition = checkwidth,
  }
}

insert_blank_line_at_right()

insert_right{
  LineInfo = {
    provider = 'LineColumn',
    -- separator = ' ',
    separator = ' ',
    separator_highlight = {colors.blue,colors.line_bg},
    highlight = {colors.blue,colors.line_bg},
        condition = checkwidth,
  }
}

insert_right{
  PerCent = {
    provider = 'LinePercent',
    separator = ' ',
    separator_highlight = {colors.green,colors.line_bg},
    highlight = {colors.green, colors.line_bg,'bold'},
        condition = checkwidth,
  }
}

insert_right{
  Encode = {
    provider = 'FileEncode',
    separator = '',
    separator_highlight = {colors.yellow,colors.line_bg},
    highlight = {colors.yellow, colors.line_bg,'bold'},
        condition = checkwidth,
  }
}

insert_blank_line_at_right()

insert_right{
  Separa = {
    provider = function() return ' ' end,
    highlight = {colors.line_bg, },
  }
}
EOF

lua << EOF
require('nvim-web-devicons').setup {
 default = true;
}
EOF
```
</details>

![image](https://user-images.githubusercontent.com/78810647/130639433-d1acf400-36cf-4dde-87f5-e97ea9efb4e7.png)

**Comment by Avimitin** [comment]
Can you try to replace the variable `vim_mode` with a hardcoded string. For example: `vim.api.nvim_command('hi GalaxyViMode guifg='..mode_color["n"]` or changes the `vim_mode` variable to a hardcode value like `local vim_mode = "n"`.

Also you can try to print the value of the `vim_mode` variable. See if it is producing a null value. Add this line below the `vim_mode` assignment: `print ("Debug: vim_mode="..vim.fn.mode())`

It seems like the table are returning nil value because of some invalid key, so I guess the `vim.fn.mode()` function is giving  invalid value back.

**Comment by Avimitin** [comment]
Cuz I am using neovim 0.6.0 I can't reproduce your problem.

**Comment by xfzv** [comment]
> Can you try to replace the variable `vim_mode` with a hardcoded string. For example: `vim.api.nvim_command('hi GalaxyViMode guifg='..mode_color["n"]` or changes the `vim_mode` variable to a hardcode value like `local vim_mode = "n"`.

The error no longer appears if replacing

```lua
-- vim.api.nvim_command('hi GalaxyViMode guifg='..mode_color[vim_mode])
vim.api.nvim_command('hi GalaxyViMode guifg='..mode_color["n"])
```

However, the Vim mode section of the bar disappears as soon as I enter visual block mode. And of course, all modes are using the color assigned to the normal mode.

Replacing

```lua
-- local vim_mode = vim.fn.mode()
local vim_mode = "n"
```

gets rid of the error too. The vim mode section only prints `N` when entering visual block mode, without any icon.

> Also you can try to print the value of the `vim_mode` variable. See if it is producing a null value. Add this line below the `vim_mode` assignment: `print ("Debug: vim_mode="..vim.fn.mode())`

With:

```lua
local vim_mode = vim.fn.mode()
vim.api.nvim_command('hi GalaxyViMode guifg='..mode_color[vim_mode])
```

It prints `Debug: vim_mode=<letter>` (n for normal, i for insert, v for visual, r for replace...) and the initial error is printed instead of the `Debug` line as soon as I enter visual block mode.

With:

```lua
local vim_mode = vim.fn.mode()
vim.api.nvim_command('hi GalaxyViMode guifg='..mode_color["n"])

-- or

local vim_mode = "n"
vim.api.nvim_command('hi GalaxyViMode guifg='..mode_color[vim_mode])

-- or

local vim_mode = "n"
vim.api.nvim_command('hi GalaxyViMode guifg='..mode_color["n"])
```

This is what I'm getting when entering visual block mode:
![image](https://user-images.githubusercontent.com/78810647/130758557-e830defa-e3dd-4472-9121-8688cbe3ea6a.png)

**Comment by Avimitin** [comment]
Ok I know what's happening, the `^V` character may be escaped by the neovim. So although it represents `^V` but the init.vim changed it to some actual key character which is not a string can be read by human and Lua. The table can't recognize the key and returning an empty value because there is not value associate with the key after escaped. I think it's a vim script-specific error.  

So my suggestion is that use a Lua file to store your Lua setting, and then source it in the init.vim file. This might help you keep the `^V` character not escaped it. By the way, this could help you maintain your configuration file. Split your configuration into small files can improve readability and make the config easier to maintain when your neovim configuration growing larger.

**Comment by xfzv** [comment]
Good catch, I finally managed to fix the issue.

For some reasons, pasting the whole lua code within konsole (KDE Plasma terminal) didn't properly include these characters:

https://github.com/Avimitin/nerd-galaxyline/blob/a0949262a3ea9421a7cc43068a2d7c3dfdd3fcba/lua/nerd-galaxyline.lua#L166
https://github.com/Avimitin/nerd-galaxyline/blob/a0949262a3ea9421a7cc43068a2d7c3dfdd3fcba/lua/nerd-galaxyline.lua#L181

I used Kate instead, all good now. It works fine in my `init.vim` file and sourcing the whole content from a dedicated lua file works too.

Thanks again for your time!
